controller/Review.php

```diff
@@ -189,47 +214,73 @@ static function (array $dataEntry): bool {
     {
         /** @var ResultServerService $resultServerService */
         $resultServerService = $this->getServiceLocator()->get(ResultServerService::SERVICE_ID);
-        /** @var \taoResultServer_models_classes_ReadableResultStorage $implementation */
+        /** @var taoResultServer_models_classes_ReadableResultStorage $implementation */
         $implementation = $resultServerService->getResultStorage();
 
         $testTaker = new core_kernel_users_GenerisUser($this->getResource($implementation->getTestTaker($resultId)));
         $lang = $testTaker->getPropertyValues(GenerisRdf::PROPERTY_USER_DEFLG);
 
-        return empty($lang) ? DEFAULT_LANG : (string)current($lang);
+        return empty($lang) ? DEFAULT_LANG : (string) current($lang);
     }
 
+    /**
+     * @throws LtiVariableMissingException
+     * @throws common_exception_NotFound
+     * @throws common_exception_Unauthorized
+     */
     protected function checkPermissions(string $serviceCallId): void
     {
-        try {
-            $execution = $this->getDeliveryExecutionFinderService()->findDeliveryExecution(
-                $this->ltiSession->getLaunchData()
-            );
-        } catch (common_Exception $e) {
-            throw new common_exception_Unauthorized($e->getMessage());
-        }
-        if ($serviceCallId !== $execution->getIdentifier()) {
+        $execution = $this->getDeliveryExecutionManagerService()->getDeliveryExecutionById($serviceCallId);
+        $userId = $this->getUserId();
+
+        if ($execution->getUserIdentifier() !== $userId) {
             throw new common_exception_Unauthorized($serviceCallId);
         }
     }
 
     private function getDeliveryExecutionFinderService(): DeliveryExecutionFinderService
     {
-        /** @noinspection PhpIncompatibleReturnTypeInspection */
-        return $this->getServiceLocator()->get(DeliveryExecutionFinderService::SERVICE_ID);
+        return $this->getPsrContainer()->get(DeliveryExecutionFinderService::SERVICE_ID);
     }
 
     private function getDeliveryExecutionManagerService(): DeliveryExecutionManagerService
     {
-        /** @noinspection PhpIncompatibleReturnTypeInspection */
-        return $this->getServiceLocator()->get(DeliveryExecutionManagerService::SERVICE_ID);
+        return $this->getPsrContainer()->get(DeliveryExecutionManagerService::SERVICE_ID);
+    }
+
+    private function getDefaultUrlService(): DefaultUrlService
+    {
+        return $this->getPsrContainer()->get(DefaultUrlService::SERVICE_ID);
+    }
+
+    private function getQtiRunnerInitDataBuilderFactory(): QtiRunnerInitDataBuilderFactory
+    {
+        return $this->getPsrContainer()->get(QtiRunnerInitDataBuilderFactory::SERVICE_ID);
+    }
+
+    private function getDeliveryId(): ?string
+    {
+        return $this->getPsrRequest()->getQueryParams()['delivery'] ?? null;
+    }
+
+    /**
+     * @throws LtiVariableMissingException
+     */
+    public function getUserId(): string
```

Comment:
If I remember well we had an internal coding convention which states about ordering methods by their access level

Reply:
it's reasonable but i think this method must be private at all

Reply:
Then why it's public? 😊

Reply:
this extract is smeared in 2 different methods of this controller